framework/gui.cpp

```diff
@@ -1245,4 +1245,26 @@ void Drawer::text(const char *formatstr, ...)
 	va_end(args);
 }
 
+bool Drawer::color_picker(const char *caption, const char channel_count, float *color, uint16_t width, ImGuiColorEditFlags flags)
+{
+	assert((channel_count == 3 || channel_count == 4) && "Channel count value must be 3 or 4");
```

Comment:
I felt bad about making my method the first template in the Drawer class if it wasn't necessary. Instead, I created two overloaded functions for different sizes of the array. I think this solution is fine.

Reply:
Well, instead of four functions, each one doing more or less the same, I would prefer a templatized approach, like this:
```
	enum class ColorOp
	{
		Edit,
		Pick
	};

	template <ColorOp OP, size_t N>
	bool color_op(const std::string & caption, std::array<float, N> &color, float width = 0.0f, ImGuiColorEditFlags flags = 0)
	{
		static_assert((N == 3) || (N == 4), "booh");

		ImGui::PushItemWidth(width);
		bool res = color_op_impl<OP, N>(caption.c_str(), color.data(), flags);
		ImGui::PopItemWidth();
		if (res)
			dirty = true;
		return res;
	}
```
With some private implementation functions (unfortunately, we're at C++14, not C++17, where we could use some `constexpr if` instead) like that:
```
	template <ColorOp OP, size_t N>
	bool color_op_impl(const char *caption, float *colors, ImGuiColorEditFlags flags)
	{
		assert(false);
	}

	template <>
	bool color_op_impl<ColorOp::Edit, 3>(const char *caption, float *colors, ImGuiColorEditFlags flags)
	{
		return ImGui::ColorEdit3(caption, colors, flags);
	}

	template <>
	bool color_op_impl<ColorOp::Edit, 4>(const char *caption, float *colors, ImGuiColorEditFlags flags)
	{
		return ImGui::ColorEdit4(caption, colors, flags);
	}

	template <>
	bool color_op_impl<ColorOp::Pick, 3>(const char *caption, float *colors, ImGuiColorEditFlags flags)
	{
		return ImGui::ColorPicker3(caption, colors, flags);
	}

	template <>
	bool color_op_impl<ColorOp::Pick, 4>(const char *caption, float *colors, ImGuiColorEditFlags flags)
	{
		return ImGui::ColorPicker4(caption, colors, flags);
	}
```
But, of course, your solution works as well, and if nobody else objects, I'd be ok as well.
(As there's no vulkan stuff with this set of functions, we could even move them out to some place, where they could be called by both, vkb::Drawer and vkb::HPPDrawer. But that might be a different story.)

Reply:
After giving it a second thought, I really liked your idea. I changed the code according to it.